### PR TITLE
fix(grpo): fail fast when continuing from a non-trainable PEFT adapter (silent no-op)

### LINKED ISSUE #6700: GRPOTrainer silently no-ops (empty optimizer) when continuing from a PeftModel loaded with is_trainable=False and beta != 0
# GRPOTrainer silently no-ops (empty optimizer) when continuing from a PeftModel loaded with `is_trainable=False` and `beta != 0`

## Problem summary

`GRPOTrainer` silently trains nothing when it is given a `PeftModel` that was loaded with `PeftModel.from_pretrained(..., is_trainable=False)` and `beta != 0`:

- The auto-created `ref` adapter (used for the KL term) inherits `inference_mode=True` from the default adapter's config.
- PEFT's `inject_adapter` then also freezes the active ("default") adapter, so after trainer init there are **zero trainable parameters** and the optimizer is created **empty**.
- During training, `use_adapter()` temporarily flips `requires_grad`, so gradients accumulate, `grad_norm` is non-zero, and loss/reward vary from step to step (each step samples different prompts). The run completes without any error or warning, and the saved adapter is **byte-identical (md5)** to the initial adapter.

We hit this in production (100-step GRPO run, complete wandb curves, everything looked healthy) and only noticed after hashing the saved weights.

## Why `is_trainable=True` in #3031 is NOT a sufficient fix

Loading the adapter with `is_trainable=True` does unblock training (verified: same script, optimizer goes from 0 to 12 params and the saved adapter changes). But it does not resolve the underlying defect:

1. **The trap stays silent for everyone else.** The default usage (`PeftModel.from_pretrained(...)` with no `is_trainable`) still silently no-ops with healthy-looking metrics; nothing warns or errors, so any future user hits it again.
2. **It cannot change the LoRA configuration.** Continuing with a different `r`, `target_modules`, or a new `peft_config` requires merging first: TRL raises `ValueError` when a `PeftModel` and a `peft_config` are passed together ("Please first merge and unload the existing adapter...").
3. **The `ref`-adapter mechanism remains fragile.** Whether `ref` stays frozen depends on current PEFT behavior, not on an explicit TRL guarantee.
4. **The default behavior is a footgun.** Until the framework either raises or auto-handles trainability, users following the "continue from a fine-tuned adapter" flow keep getting burned silently.

A fix (fail fast with a clear error + explicitly freeze `ref`) is prepared in the follow-up PR.

## Reproduction

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
from peft import LoraConfig, get_peft_model, PeftModel
from datasets import Dataset
from trl import GRPOConfig, GRPOTrainer

model_id = "sshleifer/tiny-gpt2"

tok = AutoTokenizer.from_pretrained(model_id)
tok.pad_token = tok.eos_token
tok.chat_template = (
    "{% for message in messages %}{{ message['role'] }}: "
    "{{ message['content'] }}{{ '\\n' if not loop.last else '' }}{% endfor %}"
)

lora = LoraConfig(r=8, lora_alpha=16, target_modules=["c_attn", "c_proj"], task_type="CAUSAL_LM")

# 0) create a dummy "already fine-tuned" adapter to continue from
sft = get_peft_model(AutoModelForCausalLM.from_pretrained(model_id), lora)
sft.save_pretrained("./sft_adapter")

# 1) load it WITHOUT is_trainable=True (this is the default)
model = PeftModel.from_pretrained(AutoModelForCausalLM.from_pretrained(model_id), "./sft_adapter")

train = Dataset.from_list([
    {"prompt": "What is 2+2?", "answer": "4"},
    {"prompt": "What is 3+4?", "answer": "7"},
])


def reward_fn(completions, **kwargs):
    return [(len(set(c)) % 3) / 2.0 for c in completions]  # non-uniform rewards


trainer = GRPOTrainer(
    model=model,
    processing_class=tok,
    reward_funcs=[reward_fn],
    train_dataset=train,
    args=GRPOConfig(
        output_dir="./out",
        beta=0.01,  # beta != 0 triggers the buggy path
        per_device_train_batch_size=1,
        generation_batch_size=4,
        num_generations=4,
        max_completion_length=64,
        max_steps=4,
        logging_steps=1,
        save_steps=4,
        report_to="none",
        seed=42,
    ),
)

print("trainable params after trainer init:",
      sum(1 for p in trainer.model.parameters() if p.requires_grad))

trainer.train()

print("optimizer param count (captured at creation):",
      sum(len(g["params"]) for g in trainer.optimizer.param_groups))
```

outputs:

```text
trainable params after trainer init: 0
optimizer param count (captured at creation): 0
{'loss': '-0.26', 'grad_norm': '0.001', 'reward': '0.625', ...}   # metrics look alive
```

And the saved adapter (`out/checkpoint-4/adapter_model.safetensors`) is **byte-identical (md5)** to `sft_adapter/adapter_model.safetensors`.

Control: loading with `PeftModel.from_pretrained(..., is_trainable=True)` gives `optimizer param count: 12` and a saved adapter that differs from the initial one — the exact same script trains for real.

Full runnable reproduction (CPU only, ~2 min): https://www.kaggle.com/code/meredith10pi/trl-1-9-2-ref-adapter-silent-no-op-repro

## System Info

```text
trl==1.9.2
peft==0.20.0
transformers==5.14.1
torch==2.9.0+cu128 (also reproduced with torch==2.10.0+cpu)
accelerate==1.14.0
datasets==5.0.1
Python 3.10/3.12, Linux (reproduced on a single CPU process and on 2x RTX 5090 with accelerate multi-GPU)
```

## Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by Meredith2328**
> [!NOTE]
> **Clarified scope:** The `ref` adapter is not the root cause, and changing PEFT trainability is not the goal. This PR is about preventing a misleading no-op: after accepting a model with zero trainable parameters, the trainer can complete with apparently normal gradients and metrics while the optimizer updates nothing and the saved adapter remains unchanged. The useful change is the diagnostic fail-fast; the ref-specific explanation and defensive freeze can be removed.

# What does this PR do?

`GRPOTrainer` currently **silently no-ops** when it is given a `PeftModel` that was loaded with `PeftModel.from_pretrained(..., is_trainable=False)` and `beta != 0`:

- The auto-created `ref` adapter (KL reference) inherits `inference_mode=True` from the default adapter's config, and PEFT's `inject_adapter` then freezes the active ("default") adapter too, so the optimizer is created with **zero parameters**.
- During each step, `use_adapter()` temporarily flips `requires_grad`, so loss / reward / grad_norm all look alive; the training completes without any warning and the saved adapter is **byte-identical** to the initial adapter.

This PR:

1. explicitly freezes the auto-created `ref` adapter (`set_requires_grad("ref", False)`) so the reference policy can never drift (defensive; PEFT 0.20.0 housekeeping already freezes it today);
2. adds a fail-fast check after the PEFT setup: if the model has no trainable parameters, raise a clear `ValueError` telling the user to load the adapter with `is_trainable=True` or to `merge_and_unload()` it and pass a `peft_config` — instead of silently training nothing;
3. adds regression tests for both the error path and the working `is_trainable=True` path.

The trainability check also covers the `beta == 0` case, which silently no-ops in the same way (that path never enters the `ref`-adapter branch).

Fixes #6700

## Verification

- Minimal CPU reproduction (public notebook): https://www.kaggle.com/code/meredith10pi/trl-1-9-2-ref-adapter-silent-no-op-repro
- With the fix applied (trl 1.9.2 + peft 0.20.0): `is_trainable=False + beta!=0` raises `ValueError`; `is_trainable=True + beta!=0` trains normally (optimizer non-empty, saved weights change); `is_trainable=False + beta==0` also raises; `merge_and_unload + peft_config` and `is_trainable=True + beta==0` are unaffected.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. — https://github.com/huggingface/trl/issues/6700
- [x] Did you make sure to update the documentation with your changes? — Inline comments updated; behavior change documented in the error message.
- [x] Did you write any new necessary tests? — `TestGRPOTrainerPEFTContinuation` in `tests/test_grpo_trainer.py`.

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. `albertvillanova` has been the maintainer most involved in the PEFT/ref-adapter code paths (see #3031, #5222, #5673) and may be interested.

---

> [!NOTE]
> **Medium Risk**
> Changes trainer initialization for all PEFT continuation paths across four trainers; previously broken setups now error instead of silently not training, which is safer but may break scripts that relied on the old behavior.
> 
> **Overview**
> Fixes a **silent no-op** when continuing RLHF-style training from a `PeftModel` loaded via `PeftModel.from_pretrained` with the default `is_trainable=False`: PEFT can leave the active adapter frozen and the optimizer empty while metrics still look healthy.
> 
> **DPO**, **GRPO**, **KTO**, and **RLOO** trainers now explicitly freeze the auto-created `ref` adapter (`set_requires_grad("ref", False)` on PEFT ≥ 0.18.0) and **raise `ValueError`** during init if the model has no trainable parameters and no new `peft_config` was passed, with guidance to use `is_trainable=True` or `merge_and_unload()` plus `peft_config`.
> 
> Adds **`TestGRPOTrainerPEFTContinuation`** regression tests: expect failure when the adapter is non-trainable (with `beta != 0`), and assert trainable parameters exist when `is_trainable=True`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e8ad59a11d606f9f52b92fa5d9c2114fad25f28d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: tests/test_grpo_trainer.py [hunk 1]
```diff
@@ -3699,6 +3699,78 @@ def test_single_reward_model_with_single_processing_class(self):
 
 
 @require_vision
+class TestGRPOTrainerPEFTContinuation(TrlTestCase):
```
[the remark this answers is not in the document]
Comment (Meredith2328) [reply tests/test_grpo_trainer.py:3701]: Fixed in 70bf69f6: the new \TestGRPOTrainerPEFTContinuation\ class is now placed before \@require_vision\, so the decorator stays on \TestGRPOTrainerVLM\ and the PEFT regression tests no longer inherit the vision skip marker.

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -474,6 +474,27 @@ def __init__(
...
+
+        if is_peft_model(model) and peft_config is None and not any(
+            p.requires_grad for p in model.parameters()
+        ):
+            raise ValueError(
+                "GRPOTrainer found no trainable parameters in the PEFT model. "
+                "If you loaded an existing adapter with `PeftModel.from_pretrained`, "
+                "pass `is_trainable=True`, or `merge_and_unload()` it and pass a "
+                "`peft_config` to the trainer instead."
+            )
```
[the remark this answers is not in the document]
Comment (Meredith2328) [reply trl/trainer/grpo_trainer.py:497]: Fixed in 70bf69f6: the same freeze + trainability fail-fast check is now mirrored to \RLOOTrainer\, \DPOTrainer\ and \KTOTrainer\, which share the identical \dd_adapter(\ref\, default_config)\ block (verified they all import \peft\ and \Version\).

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -474,6 +474,27 @@ def __init__(
...
+                # inherits `inference_mode` from the default adapter's config:
+                # when the model was loaded with `is_trainable=False` (the
+                # default of `PeftModel.from_pretrained`), PEFT also freezes the
+                # active "default" adapter here, so the optimizer would be
+                # created empty and GRPO would silently train nothing. Freeze
+                # the reference explicitly (defensive: PEFT 0.20.0 housekeeping
+                # already freezes it) and rely on the trainability check below
+                # to catch the empty-optimizer case.
+                if not model.peft_config["ref"].is_prompt_learning:
+                    model.set_requires_grad("ref", requires_grad=False)
```
[the remark this answers is not in the document]
Comment (Meredith2328) [reply trl/trainer/grpo_trainer.py:487]: Fixed in 70bf69f6: \set_requires_grad(\ref\, False)\ is now guarded by \Version(peft.__version__) >= Version(\0.18.0\)\, so older supported PEFT versions fall back to the previous behavior (ref stays frozen via PEFT housekeeping) and the trainability check still raises the intended ValueError.

**Comment by Meredith2328** [comment]
Hi @albertvillanova — a small follow-up on this reproducible silent training no-op.

The PR's existing CI was triggered, but both runs are still `action_required` and contain no jobs, so the test matrix and PR-docs build have not actually run yet:

- [Tests](https://github.com/huggingface/trl/actions/runs/31387458393)
- [Build PR Documentation](https://github.com/huggingface/trl/actions/runs/31387458827)

Could a maintainer approve those workflows when convenient? The existing path filters already cover the changed `trl/**.py` and `tests/**.py` files, so I have deliberately not changed CI configuration. The GRPO regression test covers both the failing default `is_trainable=False` continuation and the working `is_trainable=True` control. If you would prefer the change to stay GRPO-only, or want analogous tests for the other affected trainers, I am happy to adjust it.

**Comment by qgallouedec** [comment]
Closing, see https://github.com/huggingface/trl/issues/6700#issuecomment-5736194915
